oauth2_http/javatests/com/google/auth/oauth2/ITWorkloadIdentityFederationTest.java

```diff
@@ -0,0 +1,261 @@
+    String idToken = generateGoogleIdToken(OIDC_AUDIENCE);
+
+    File file =
+        File.createTempFile(
+            "ITWorkloadIdentityFederation", /* suffix= */ null, /* directory= */ null);
+    file.deleteOnExit();
+    OAuth2Utils.writeInputStreamToFile(
+        new ByteArrayInputStream(idToken.getBytes(StandardCharsets.UTF_8)), file.getAbsolutePath());
+
+    GenericJson config = new GenericJson();
+    config.put("type", "external_account");
+    config.put("audience", OIDC_AUDIENCE);
+    config.put("subject_token_type", "urn:ietf:params:oauth:token-type:jwt");
+    config.put("token_url", "https://sts.googleapis.com/v1beta/token");
+    config.put(
+        "service_account_impersonation_url",
+        String.format(
+            "https://iamcredentials.googleapis.com/v1/projects/-/serviceAccounts/%s:generateAccessToken",
+            clientEmail));
+
+    GenericJson credentialSource = new GenericJson();
+    credentialSource.put("file", file.getAbsolutePath());
+    config.put("credential_source", credentialSource);
+
+    return config;
+  }
+
+  private GenericJson buildAwsCredentialConfig() {
+    GenericJson config = new GenericJson();
+    config.put("type", "external_account");
+    config.put("audience", AWS_AUDIENCE);
+    config.put("subject_token_type", "urn:ietf:params:aws:token-type:aws4_request");
+    config.put("token_url", "https://sts.googleapis.com/v1beta/token");
+    config.put(
+        "service_account_impersonation_url",
+        String.format(
+            "https://iamcredentials.googleapis.com/v1/projects/-/serviceAccounts/%s:generateAccessToken",
+            clientEmail));
+
+    GenericJson credentialSource = new GenericJson();
+    credentialSource.put("environment_id", "aws1");
+    credentialSource.put(
+        "regional_cred_verification_url",
+        "https://sts.{region}.amazonaws.com?Action=GetCallerIdentity&Version=2011-06-15");
+    config.put("credential_source", credentialSource);
+
+    return config;
+  }
+
+  private void callGcs(GoogleCredentials credentials) throws IOException {
+    String bucketName = "byoid-it-bucket";
+    String url = "https://storage.googleapis.com/storage/v1/b/" + bucketName;
+
+    HttpCredentialsAdapter credentialsAdapter = new HttpCredentialsAdapter(credentials);
+    HttpRequestFactory requestFactory =
+        new NetHttpTransport().createRequestFactory(credentialsAdapter);
+    HttpRequest request = requestFactory.buildGetRequest(new GenericUrl(url));
+
+    JsonObjectParser parser = new JsonObjectParser(GsonFactory.getDefaultInstance());
+    request.setParser(parser);
+
+    HttpResponse response = request.execute();
+    assertTrue(response.isSuccessStatusCode());
+  }
+
+  /**
+   * Generates a Google ID token using the iamcredentials generateIdToken API.
+   * https://cloud.google.com/iam/docs/creating-short-lived-service-account-credentials#sa-credentials-oidc
+   */
+  private String generateGoogleIdToken(String audience) throws IOException {
```

Comment:
For ID Tokens, we do have `IdTokenCredentials` which abstracts the token exchange from credentials that are `IdTokenProvider`s so you con't have to rewrite all this code.

Reply:
For ServiceAccountCredentials#idTokenWithAudience, the default tokenServerUri returns an ID token with a 'azp' claim that AWS doesn't like. If I override it to use the generateIdToken API, I get a 401 (I'm guessing because it uses a JWT, not entirely sure why)